pkg/cloud/rgraph/rnode/action_update.go

```diff
@@ -18,34 +18,122 @@ package rnode
+	ops GenericOps[GA, Alpha, Beta],
+	got, want Node,
+	resource api.Resource[GA, Alpha, Beta],
+) ([]exec.Action, error) {
+	preEvents, err := updatePreconditions(got, want)
+	if err != nil {
+		return nil, err
+	}
+	postEvents := postUpdateActionEvents(got, want)
+	return []exec.Action{
+		newGenericUpdateAction(preEvents, ops, want.ID(), resource, postEvents),
+	}, nil
+}
+
+func newGenericUpdateAction[GA any, Alpha any, Beta any](
+	want exec.EventList,
+	ops GenericOps[GA, Alpha, Beta],
+	id *cloud.ResourceID,
+	resource api.Resource[GA, Alpha, Beta],
+	postEvents exec.EventList,
+) *genericUpdateAction[GA, Alpha, Beta] {
+	return &genericUpdateAction[GA, Alpha, Beta]{
+		ActionBase: exec.ActionBase{Want: want},
+		ops:        ops,
+		id:         id,
+		resource:   resource,
+		postEvents: postEvents,
+	}
+}
+
 type genericUpdateAction[GA any, Alpha any, Beta any] struct {
 	exec.ActionBase
+	ops        GenericOps[GA, Alpha, Beta]
+	id         *cloud.ResourceID
+	resource   api.Resource[GA, Alpha, Beta]
+	postEvents exec.EventList
+
+	start, end time.Time
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) Run(
 	ctx context.Context,
 	c cloud.Cloud,
 ) (exec.EventList, error) {
-	return nil, nil
+	a.start = time.Now()
+	err := a.ops.UpdateFuncs(c).Do(ctx, "", a.id, a.resource)
+	a.end = time.Now()
+
+	// Emit DropReference events for removed references.
+	return a.postEvents, err
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) DryRun() exec.EventList {
-	return nil
+	// Emit DropReference events for removed references.
+	return a.postEvents
 }
 
 func (a *genericUpdateAction[GA, Alpha, Beta]) String() string {
-	return "GenericUpdateAction TODO"
+	return fmt.Sprintf("GenericUpdateAction(%v)", a.id)
+}
+
+func (a *genericUpdateAction[GA, Alpha, Beta]) Metadata() *exec.ActionMetadata {
+	return &exec.ActionMetadata{
+		Name:    fmt.Sprintf("GenericUpdateAction(%s)", a.id),
+		Type:    exec.ActionTypeUpdate,
+		Summary: fmt.Sprintf("Update %s", a.id),
+	}
 }
 
-func updatePreconditions(got, want Node) exec.EventList {
+func updatePreconditions(got, want Node) (exec.EventList, error) {
 	// Update can only occur if the resource Exists TODO: is there a case where
 	// the ambient signal for existance from Update op collides with a
 	// reference to it?
-	return nil // TODO: finish me
+	if got.State() != NodeExists || want.State() != NodeExists {
+		return nil, fmt.Errorf("node for update does not exist")
+	}
+
+	outRefs := want.OutRefs()
+	var events exec.EventList
+	// Condition: references must exist before update.
+	for _, ref := range outRefs {
+		events = append(events, exec.NewExistsEvent(ref.To))
+	}
+	return events, nil
+}
+
+func postUpdateActionEvents(got, want Node) exec.EventList {
+	wantOutRefs := want.OutRefs()
+	gotOutRefs := got.OutRefs()
+
+	diff := make(map[meta.Key]ResourceRef)
+	for _, r := range gotOutRefs {
+		diff[*r.To.Key] = r
+	}
+
+	for _, wantRef := range wantOutRefs {
```

Comment:
What if we made the map of the wantRefs and then constructed the droprefevents directly?

Seems like the set of dropped = got - want.

then the logic would be 

```
wantMap := ...
for _, gotRef := range gotOutRefs {
  if gotRef not in wantMap { append }
}
```

Reply:
would not have to delete map entries and loop again.

Reply:
done